moveit_ros/planning_interface/test/plan_using_subframes.cpp

```diff
@@ -0,0 +1,299 @@
+
+  box.subframe_names[4] = "side";
+  box.subframe_poses[4].position.x = .0;
+  box.subframe_poses[4].position.y = .0;
+  box.subframe_poses[4].position.z = -.01 + z_offset_box;
+  orientation.setRPY(0, (180.0 / 180.0 * M_PI), 0);
+  box.subframe_poses[4].orientation = tf2::toMsg(orientation);
+
+  // Next, define the cylinder
+  moveit_msgs::CollisionObject cylinder;
+  cylinder.id = "cylinder";
+  cylinder.header.frame_id = "panda_hand";
+  cylinder.primitives.resize(1);
+  cylinder.primitive_poses.resize(1);
+  cylinder.primitives[0].type = box.primitives[0].CYLINDER;
+  cylinder.primitives[0].dimensions.resize(2);
+  cylinder.primitives[0].dimensions[0] = 0.06;   // height (along x)
+  cylinder.primitives[0].dimensions[1] = 0.005;  // radius
+  cylinder.primitive_poses[0].position.x = 0.0;
+  cylinder.primitive_poses[0].position.y = 0.0;
+  cylinder.primitive_poses[0].position.z = 0.0 + z_offset_cylinder;
+  orientation.setRPY(0, (90.0 / 180.0 * M_PI), 0);
+  cylinder.primitive_poses[0].orientation = tf2::toMsg(orientation);
+
+  cylinder.subframe_poses.resize(1);
+  cylinder.subframe_names.resize(1);
+  cylinder.subframe_names[0] = "tip";
+  cylinder.subframe_poses[0].position.x = 0.03;
+  cylinder.subframe_poses[0].position.y = 0.0;
+  cylinder.subframe_poses[0].position.z = 0.0 + z_offset_cylinder;
+  orientation.setRPY(0, (90.0 / 180.0 * M_PI), 0);
+  cylinder.subframe_poses[0].orientation = tf2::toMsg(orientation);
+
+  // Lastly, the objects are published to the PlanningScene. In this tutorial, we publish a box and a cylinder.
+  box.operation = moveit_msgs::CollisionObject::ADD;
+  cylinder.operation = moveit_msgs::CollisionObject::ADD;
+  std::vector<moveit_msgs::CollisionObject> collision_objects = { box, cylinder };
+  planning_scene_interface.applyCollisionObjects(collision_objects);
+}
+
+void poseMsgToEigen(const geometry_msgs::Pose& m, Eigen::Affine3d& e)
+{
+  e = Eigen::Translation3d(m.position.x, m.position.y, m.position.z) *
+      Eigen::Quaterniond(m.orientation.w, m.orientation.x, m.orientation.y, m.orientation.z);
+}
+
+void testNearSubframe(const std::string& subframe_name,
+                      moveit::planning_interface::PlanningSceneInterface* planning_scene_interface,
+                      moveit::planning_interface::MoveGroupInterface* group)
+{
+  geometry_msgs::Pose tip_in_hand_msg, hand_in_world_msg, tip_in_world_msg, box_subframe_msg;
+  Eigen::Affine3d tip_in_hand, tip_in_world, hand_in_word, box_subframe_in_world;
+
+  auto attached_objects = planning_scene_interface->getAttachedObjects();
+  tip_in_hand_msg = attached_objects["cylinder"].object.subframe_poses[0];
+  hand_in_world_msg = group->getCurrentPose("panda_hand").pose;
+  poseMsgToEigen(tip_in_hand_msg, tip_in_hand);
+  poseMsgToEigen(hand_in_world_msg, hand_in_word);
+  tip_in_world = hand_in_word * tip_in_hand;
+
+  auto box = planning_scene_interface->getObjects()["box"];
+  auto box_subframe_names = box.subframe_names;
```

Comment:
We discourage usage of auto in MoveIt, as it hurts readability. Is this a string? 

> However, auto is not an expression highly readable. Please explicitly specify the variable type, if it doesn’t become clear immediately from the context.

Reply:
https://moveit.ros.org/documentation/contributing/code/